### PR TITLE
Use unblocking io

### PR DESCRIPTION
**Pull request by ecanzonieri**
Use select to implement unblocking IO when sending broker aware requests

### CODE REVIEW

File: kafka/client.py [hunk 1]
```diff
@@ -223,27 +228,34 @@ def _send_broker_aware_request(self, payloads, encoder_fn, decoder_fn):
...
-                    response = conn.recv(requestId)
-                except ConnectionError as e:
-                    broker_failures.append(broker)
-                    log.warning('ConnectionError attempting to receive a '
-                                'response to request %s from server %s: %s',
-                                requestId, broker, e)
+        conn = None
+        while connections_by_socket:
+            sockets = connections_by_socket.keys()
+            rlist, _, _ = select.select(sockets, [], [], None)
```
Comment (yyejun) [inline kafka/client.py]: If the read socket has exception, will this select block because 3rd parameter is []?

**Comment by yyejun** [comment]
overall I feel the benefit of this change is small. It barely saves the time to handle the first connection, assuming the majority time is spent in the recv call. Because eventually all requests need to be processed.. The actual receiving still in serial, after first request processed, it's very likely all other socket are all ready to be received, so the select on 2nd request, mostly like all socket are ready.

File: kafka/client.py (hunk 1 above)
Comment (ecanzonieri) [inline kafka/client.py]: Sockets exceptions are marked in the readfds, so the affected socket should be returned as read available and the subsequent recv will raise an exception.

**Comment by ecanzonieri** [comment]
The improvements in our larger kafka cluster is notable. For watermark fetch the latency goes down from 1.70s to 0.35s, while consumer offset fetch goes down from 12.5 seconds to 1.20s.

**Comment by yyejun** [comment]
lgtm

**Pull request merged**